internal-services/catalog/iib-add-fbc-fragment-to-index-image-task.yaml

```diff
@@ -83,49 +83,75 @@ spec:
       script: |
         #!/usr/bin/env bash
         #
+        # shellcheck disable=SC2207,SC2128,SC2178 # requires refactoring to fix these errors
+        # shellcheck disable=SC2005 # false positive as it thinks tekton parameters might be commands
         isfbcOptIn() {
           TMPFILE=$(mktemp)
           PYXIS_URL="https://pyxis.engineering.redhat.com/v1"
 
-          IFS="/" read REGISTRY REPO IMAGE <<< "${1}"
-          IFS=":" read IMAGE TAG <<< "${IMAGE}"
+          IFS="/" read -r REGISTRY REPO IMAGE <<< "${1}"
+          IFS=":" read -r IMAGE TAG <<< "${IMAGE}"
 
           FETCH_URL="${PYXIS_URL}/repositories/registry/${REGISTRY}/repository/${REPO}/${IMAGE}/tag/${TAG}"
 
           # strips the last "/tag" in case $TAG is not set
           [ -z "${TAG}" ] && FETCH_URL=${FETCH_URL%/tag*}
 
-          curl --negotiate -u: "${FETCH_URL}" -o $TMPFILE
+          curl --negotiate -u: "${FETCH_URL}" -o "${TMPFILE}"
           # prints "false" in case .fbc_opt_in entry is missing
-          jq -e -r '.fbc_opt_in //false' $TMPFILE && rm -f $TMPFILE
+          jq -e -r '.fbc_opt_in //false' "${TMPFILE}" && rm -f "${TMPFILE}"
+        }
+
+        # checks if there was any previous build for the same snapshot,
+        # although it should only consider resume it if is not a stage
+        # build as stage builds have no reason for being resumed.
+        check_previous_prod_build() {
+
+          user="${1}"
+          from_index="${2}"
+          fbc_fragment="${3}"
+
+          # fetch only builds in progress or completed
+          curl -s "https://iib.engineering.redhat.com/api/v1/builds?user=${user}&from_index=${from_index}" | \
+            jq -r --arg fbc_fragment "${fbc_fragment}" '.items[] |select(.fbc_fragment="$fbc_fragment")'
         }
 
         # performs kerberos authentication.
         base64 -d /mnt/service-account-secret/keytab > /tmp/keytab
 
         KRB5_TEMP_CONF=$(mktemp)
+        KRB5_PRINCIPAL=$(cat /mnt/service-account-secret/principal)
+
+        build=$(check_previous_prod_build "${KRB5_PRINCIPAL}" "$(params.fromIndex)" "$(params.fbcFragment)")
+        if [ -n "${build}" ]; then
+          echo "The fragment is already building in IIB. Resuming..."
```

Comment:
What if the build was completed?

Reply:
It will get the information of  completed and exit the step. Later, in the next step it will check the status of this build and fill up the results.

Reply:
I'm confused. This `check_previous_prod_build` part seems like the only new code to make the task idempotent. If there is a previous prod build, we say it is already building and we are resuming. Couldn't the previous prod build have already succeeded? So maybe just the comment is wrong? Because if we are resuming, why are we just calling echo then exiting without any other call?

Reply:
A think what is more confusing is  "what we have here to resume".  

In my understanding, the reasons a user would need to "retry" the build would be:

- the pipeline failed/timed out, but IIB build already finished successful on retry time;
- the pipeline failed/timed out, but IIB build is still in progress on retry time;

Only in case it is the 1st build, or a build that IIB returned  **failed**  it would need to create a new build (krb auth and IIB build call)

That said, IIB has only 3 expected status: "completed", "inprogress" ( or some similar string) and "failed".  So, the only condition it needs to continue in that  "step" is if the new function check_previous_prod_build` returns a empty string (as it ignores "failed") - unless we want to finish the pipeline and show the previous failure to the user; but I think this would for the user having a new snapshot, and FBC users use the same snapshot for different `from_index` parameters.

For "in progress" and "complete", it `exit 0` because all it needs to do is in the next "step" - the second container it spawns -, where it can run the `watch_build_state()` using the json saved in the previous step.

For the moment, the only possibility of "resume" is for prod builds, because if the RPA has the `stagedIndex` set to `true`, it has nothing to resume, because IIB does not  replace the `fromIndex` image,  so there is literally nothing to do - it won't sign and won't publish, so the only thing it can do is _run a new build_ .

Maybe I am skipping something, but anyhow the code is not properly tested. I will be doing this on Monday.

Reply:
One note:  there is a mistake indeed in the jq, which is missing specifying what it should check (non failed builds) - I missed it after trying the code, to be fixed.

For a successful IIB build, it would still need to fill up all results so the public `add-fbc-fragment` task can get the results to then publish the image - but indeed, without calling IIB for a new build.
 
What it actually does (although missing the fix on the `jq` that I  mentioned above) is in case the build is in progress or succeeded/completed, jump to the script in line 230 - which only fill up the results, without calling a new build.

The `exit 0` will exit the script , so I won`t call IIB for that purpose.

Anyhow, I am starting testing this - so if I am missing something (as it is likely) I should catch on testing.

Reply:
> What if the build succeeded last time?

@johnbieren in that case `exit 0` is called two lines below this thread, so you will never get to the "start new IIB build" curl call.

So I think it's correct, but it would be useful if we printed some more details about what we found - I think it would be good to clearly see in the logs if it was already succeeded, or even failed.

One thing that is still not clear to me is why for stage builds we always need a new build. Why can't we reuse a previous build? You say "there is nothing to do" and yet we need to run a new IIB build. Why?

Reply:
@mmalina two lines below is still in this if statement. The first line of the if says "fragment is already building, resuming." So, the comment is wrong, or the case where a build for the image is already completed is not addressed. Either way, something has to change

Reply:
@johnbieren I think the comments make it clearer that it's in progress or completed.

Reply:
AFAIK, if a build was successful and the `user`, `from_index` and `fbc_fragment` is the same (I guess these are the unique params from your comment), we should not rebuild the fragment. That is the point of idempotent